cf-agent/files_editxml.c

```diff
@@ -1147,54 +1359,48 @@ static bool DeleteTreeInNode(EvalContext *ctx, char *rawtree, xmlDocPtr doc, xml
     xmlNodePtr deletetree = NULL;
     if ((deletetree = XmlVerifyNodeInNodeSubset(treenode, docnode, a, pp)) == NULL)
     {
-        cfPS(ctx, LOG_LEVEL_VERBOSE, PROMISE_RESULT_NOOP, pp, a, "The promised tree to be deleted '%s' does NOT exist, at XPath '%s' in XML document '%s' (promise kept)",
-             rawtree, a->xml.select_xpath, edcontext->filename);
+        RecordNoChange(ctx, pp, a, "The promised tree to be deleted '%s' does not exist at XPath '%s' in XML document '%s'",
+                       rawtree, a->xml.select_xpath, edcontext->filename);
         xmlFreeNode(treenode);
         return false;
     }
 
-    if (a->transaction.action == cfa_warn)
+    if (DONTDO || a->transaction.action == cfa_warn)
     {
-        cfPS(ctx, LOG_LEVEL_WARNING, PROMISE_RESULT_WARN, pp, a,
-             "Need to delete the promised tree '%s' at XPath '%s' in XML document '%s' - but only a warning was promised",
-             rawtree, a->xml.select_xpath, edcontext->filename);
+        RecordWarning(ctx, pp, a,
+                      "Tree '%s' at XPath '%s' in XML document '%s' should be deleted",
+                      rawtree, a->xml.select_xpath, edcontext->filename);
         *result = PromiseResultUpdate(*result, PROMISE_RESULT_WARN);
         xmlFreeNode(treenode);
         xmlFreeNode(deletetree);
-        return true;
+        return false;
```

Comment:
Why was this change made? It seems intentional that the warn only would return true (success / would be edited)?

```C
    if (DeleteTreeInNode(ctx, pp->promiser, doc, docnode, &a, pp, edcontext, &result))
    {
        (edcontext->num_edits)++;
    }
```

From searching the code, it seems that in most places it doesn't update `num_edits` in the warn only case, so I guess your change is correct. Would have preferred a separate commit, maybe, but no big deal.

Reply:
Yeah, it doesn't make any change.